### PR TITLE
filter tokens by notification settings

### PR DESCRIPTION
**Pull request by MixailE**

### CODE REVIEW

File: src/helpers/runNotificationListener.ts [hunk 1]
```diff
@@ -51,13 +54,16 @@ getFeedsContract.on(
...
   ) => {
-    const feed = rootFeeds[feedId.toNumber()]
-    const title = feed && `@${generateRandomName(author)} posted at ${feed}`
-    if (!title) return
-    const body = feed && (await getIPFSContent(structToCid(metadata)))
-
-    const tokens = await getAllValidTokens()
     try {
+      const feed = rootFeeds[feedId.toNumber()]
+      const title = feed && `@${generateRandomName(author)} posted at ${feed}`
```
Comment (T-Damer) [inline src/helpers/runNotificationListener.ts:59]: `if (!feed) return`

File: src/models/NotificationSettings.ts [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+) {
+  const tokens = await TokenModel.find({
+    token: { $not: { $regex: notificationTokenRegex } },
+  })
+  if (!params) return tokens.map(({ token }) => token.toString())
+  const settings = await NotificationsSettingsModel.find(params)
+  const tokenSet = new Set(settings.map(({ token }) => token?.toString()))
+  return tokens
+    .filter(({ _id }) => !tokenSet.has(_id.toString()))
+    .map(({ token }) => token.toString())
```
Comment (T-Damer) [inline src/models/NotificationSettings.ts:76]: Feels like we can reduce number of `map`, `find` and `filter` operations to 1 or 2

File: src/controllers/notificationSettings.ts [hunk 3]
```diff
@@ -83,17 +83,14 @@ export default class SettingsController {
...
-  async getData(@Params('token') token: string, @Ctx() ctx: Context) {
+  getData(@Params('token') token: string, @Ctx() ctx: Context) {
     try {
-      const tokenRecord = await TokenModel.findOne({
-        token,
-      })
-      if (!tokenRecord) return ctx.throw(badRequest("Can't find the token"))
-
-      return findOneOrCreate(tokenRecord)
+      const sett = findOneOrCreate({ token })
```
Comment (T-Damer) [inline src/controllers/notificationSettings.ts:88]: We need to test  this

File: src/models/TokenWithSettings.ts [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+) {
+  const tokens = await TokenModel.find({
+    token: { $not: { $regex: notificationTokenRegex } },
+  })
+  if (!params) return tokens.map(({ token }) => token.toString())
+  const settings = await TokenModel.find(params)
+  const tokenSet = new Set(settings.map(({ token }) => token?.toString()))
+  return tokens
+    .filter(({ _id }) => !tokenSet.has(_id.toString()))
+    .map(({ token }) => token.toString())
```
Comment (T-Damer) [inline src/models/TokenWithSettings.ts:50]: We should simplify this code and try to run through arrays less times

**Comment by MixailE** [comment]
[Make notification tokens expirable](https://trello.com/c/SDc8Q4uT/3870-make-notification-tokens-expirable)

File: src/models/TokenWithSettings.ts [hunk 5]
```diff
@@ -36,18 +38,13 @@ export async function updateTokenWithSettings(
...
 
-export async function excludeTokensWithParams(
-  params?: Partial<TokenWithSettings>
-) {
+export async function getTokens(params?: Partial<TokenWithSettings>) {
   const tokens = await TokenModel.find({
-    token: { $not: { $regex: notificationTokenRegex } },
+    ...params,
+    expired: false,
+    token: { $regex: notificationTokenRegex },
```
Comment (T-Damer) [inline src/models/TokenWithSettings.ts:45]: Actually, why we need a regex here?

**Review by T-Damer — APPROVED** [review]
Looking good

File: src/models/TokenWithSettings.ts (hunk 5 above)
Comment (MixailE) [reply src/models/TokenWithSettings.ts]: I've removed it

**Pull request merged**